lib/Doctrine/DBAL/Driver/V3CompatStatement.php

```diff
@@ -0,0 +1,203 @@
+    {
+        return $this->stmt->fetch($fetchMode, $cursorOrientation, $cursorOffset);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function fetchAll($fetchMode = null, $fetchArgument = null, $ctorArgs = null)
+    {
+        return $this->stmt->fetchAll($fetchMode, $fetchArgument, $ctorArgs);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function fetchColumn($columnIndex = 0)
+    {
+        return $this->stmt->fetchColumn($columnIndex);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function fetchNumeric()
+    {
+        return $this->stmt->fetch(PDO::FETCH_NUM);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function fetchAssociative()
+    {
+        return $this->stmt->fetch(PDO::FETCH_ASSOC);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function fetchOne()
+    {
+        return FetchUtils::fetchOne($this);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function fetchAllNumeric(): array
+    {
+        return FetchUtils::fetchAllNumeric($this);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function fetchAllAssociative(): array
+    {
+        return FetchUtils::fetchAllAssociative($this);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function fetchAllKeyValue(): array
+    {
+        return FetchUtils::fetchAllKeyValue($this);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function fetchAllAssociativeIndexed(): array
+    {
+        return FetchUtils::fetchAllAssociativeIndexed($this);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function fetchFirstColumn(): array
+    {
+        return FetchUtils::fetchFirstColumn($this);
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @return Traversable<int,array<int,mixed>>
+     */
+    public function iterateNumeric(): Traversable
+    {
+        return FetchUtils::iterateNumeric($this);
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @return Traversable<int,array<string,mixed>>
+     */
+    public function iterateAssociative(): Traversable
+    {
+        return FetchUtils::iterateAssociative($this);
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @return Traversable<mixed,mixed>
+     */
+    public function iterateKeyValue(): Traversable
+    {
+        return FetchUtils::iterateKeyValue($this);
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @return Traversable<mixed,array<string,mixed>>
+     */
+    public function iterateAssociativeIndexed(): Traversable
+    {
+        return FetchUtils::iterateAssociativeIndexed($this);
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @return Traversable<int,mixed>
+     */
+    public function iterateColumn(): Traversable
+    {
+        return FetchUtils::iterateColumn($this);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function rowCount()
+    {
+        throw Exception::notSupported('rowCount');
```

Comment:
Doesn't it make the implementation incompatible with the old one? It is possible to count rows on the statement in 2.x.

Reply:
The class decorate the ResultStatement which does not contain this method but some subclasses implements it. Maybe a simple method_exists() check can do the job ?